File: 42-trapping-rain-water/42-trapping-rain-water.java
class Solution {
    public int trap(int[] arr) {
        
        int N=arr.length;
        int[] leftmax=new int[N];
        int[] rightmax=new int[N];
        
        int max=arr[0];
        
        for(int i=0;i<N;i++){
            max=Math.max(max,arr[i]);
            leftmax[i]=max;
        }
        max=arr[N-1];
        for(int i=N-1;i>=0;i--){
            max=Math.max(max,arr[i]);
            rightmax[i]=max;
        }
        
        int water=0;
        for(int i=0;i<N;i++){
            
            // System.out.print(leftmax[i]+" ");
            
            int height=Math.min(leftmax[i],rightmax[i])-arr[i];
            
            water+=height;
        }
        
        return water;
    }
}

File: 42-trapping-rain-water/42-trapping-rain-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    
    let N=height.length;
    let leftArr=[];
    let rightArr=[];
    let ans=0;
    leftArr[0]=height[0]
    rightArr[N-1]=height[N-1];
    
    for(let i=1;i<N;i++){
        leftArr[i]=Math.max(leftArr[i-1],height[i]);
    }
    
    for(let i=N-2;i>=0;i--){
        rightArr[i]=Math.max(rightArr[i+1],height[i]);
    }
    
    for(let i=0;i<N;i++){
        let h=(Math.min(rightArr[i],leftArr[i]))-height[i];
        ans+=h;
    }
    
    return ans;
};